### PR TITLE
Add include_once function.

### PR DESCRIPTION
**Pull request by hugot**
This allows for scripts to include other scripts that include other scripts without sourcing the same scripts multiple times.

### CODE REVIEW

**Comment by juanibiapina** [comment]
Why not do this as part of the `include` function itself? I don't see any drawbacks.

**Comment by redrock9** [comment]
@juanibiapina I'm sorry for the late reply, I have been very busy lately. I made it art of the `include` function as you proposed, didn't really see any drawbacks either.

**Comment by juanibiapina** [comment]
Why did you include so many changes in the last commit? It would be hard to come back in the future and understand what is changed there because it's a feature mixed with changes to the output plus refactors.

**Comment by hugot** [comment]
You're right, sorry it just kind of happened on auto-pilot. I'll revert the commit and make the changes less drastic this week.

**Comment by hugot** [comment]
I'm sorry for the delay. I reverted the commit and made the changes more minimal in my last commit.

File: tests/include.bats [hunk 1]
```diff
@@ -47,3 +47,33 @@ load test_helper
   assert_success
   assert_output "DONE"
 }
+
+@test "includes a file only once" {
```
Comment (juanibiapina) [inline tests/include.bats]: In this test you're only testing that `BASHER_INCLUDED_ONCE` is correctly set. The test shouldn't actually care about the internals of how `include` keeps its state. A better approach would be to use a script which outputs something and check that the output is only printed once, even if the file is included two times.

File: lib/include.sh [hunk 2]
```diff
@@ -1,8 +1,9 @@
 include() {
+  [[ "$BASHER_INCLUDED_ONCE" == *"$1/$2"* ]] && return 0
```
Comment (juanibiapina) [inline lib/include.sh]: This check can cause problems. Once example case: `user/pack` and `user/package`

Comment (juanibiapina) [reply lib/include.sh]: A possible solution could be to use an array.

**Comment by juanibiapina** [comment]
Also, I'd like it better if you could just squash everything into one commit, since this is only one conceptual change.

If you don't feel like making any of these changes, I can take over.

File: tests/include.bats (hunk 1 above)
Comment (juanibiapina) [reply tests/include.bats]: Nevermind, I saw the actual test now. We should probably just delete the check on `BASHER_INCLUDED_ONCE`

**Comment by juanibiapina** [comment]
I went ahead and did it because I had some time. I hope you don't mind.

**Comment by juanibiapina** [comment]
Actually arrays are not support on `sh` and `dash`, so if we want to maintain compatibility we'd need to find a better way using string separators like you did, but preventing name conflicts.

**Comment by kadaan** [comment]
> Actually arrays are not support on `sh` and `dash`, so if we want to maintain compatibility we'd need to find a better way using string separators like you did, but preventing name conflicts.

@juanibiapina Regarding sh and dash support.  It isn't really testable right now bats doesn't support OS X and bats-core doesn't support sh or dash.

**Comment by juanibiapina** [comment]
@kadaan I use `bats` on OS X

**Comment by kadaan** [comment]
@juanibiapina I think the issue is bats 0.4.0 doesn't work on OS X due to this long open bug: [sstephenson/bats#140](https://github.com/sstephenson/bats/issues/140)

**Comment by juanibiapina** [comment]
@kadaan It works for me on bash 4 and 5. I have no idea how to even reproduce the original bug.

I install `bats` using `basher`, not a global install or homebrew, which could make a weird kind of difference.

**Comment by kadaan** [comment]
@juanibiapina Right, but it doesn't work on bash 3.2, which is the default on OS X and I don't see that changing because of the bash license change.

**Comment by juanibiapina** [comment]
Closing. More context here: https://github.com/basherpm/basher/pull/54

**Pull request closed without merging**